### DISCUSSION TITLE
Filtering support

### DISCUSSION THREAD

**Discussion by patelprateek** [milvus-io/milvus#28135] [Q&A and General discussion]
I have few index that are about 100mill - 1 billion docs
I want to use milvus for our use cases , can you please provide some details on how we support filtering so that we can do some due diligence on if it meets our SLA requirements (any benchmarks even better). 

Can you please let me know what kind of filters are supported : 
for example 
1) country=US
2) some_threshold > 0.725
3) AND/OR operator for multiple such conditions
4) 15 < age < 25 : range
5) documents in list [ ... ] 

Do we use some inverted index for enabling these pre-filtering search  Or is this. a linear scan on multiple sharded nodes .If you can point me to some design docs or apis to understand the performance consideration and memory implcations , different kind of indexes we build on such metadata along with embedding would be very helpful

**Comment by yhmo** [comment]
Filter expressions are described here: https://milvus.io/docs/boolean.md

1. country=US  ----------------------------------------------- Supported. Example: `collection.search(expr = "country=\"US\"")`
2. some_threshold > 0.725 ------------------------------- Supported. Example: `collection.search(expr = "some_threshold > 0.725")`
3. AND/OR operator for multiple such conditions --- Supported. Example: `collection.search(expr = "a > 10 && b < 5")`
4. 15 < age < 25 : range ----------------------------------- No special Keywords for range, just use AND operator, Example: `collection.search(expr = "age > 15 && age < 25")`
5. documents in list [ ... ] ----------------------------------- Supported. Keywords "in". Example: `collection.search(expr = "documents in [1, 2, 3]")
`

In milvus, a filter is always pre-filtering search: do filtering first and then ANN search. Pre-filtering scans each row in a collection, marks some rows to be "ignored", and ANN search will ignore these rows.


You can create index for scalar fields.
`collection.create_index(field_name="xxx")`

For VARCHAR field, it creates Trie(prefix tree) index.
For numeric fields, it sorts the values as index.

**Reply by patelprateek** [reply]
@yhmo : Do we build different kind of indexers for all these fields for example an inverted index for country (us -> posting list ) , sorted index for some_threshold and age 
Form some older threads here : https://github.com/milvus-io/milvus/discussions/12927 and here https://github.com/milvus-io/milvus/discussions/12927 It seems that we full brute force filtering using these boolean predicates  
1) is this linear scan or using different kind of indexes as you mentioned (trie, sorted) . Do these indexes have to be in memory or supported on disk ?
2) Post filtering : all the valid embeddings what kind of ann algorithm is used, again during query time it doest seem plausible to create any index here but do a brute force knn . Do we build any ANN index at all ?

**Reply by patelprateek** [reply]
Another follow up i have is if we are building different indexes for different attribute fields , do we duplicate the document ids in each of the index , for example if i have two numeric fields , do we build one sorted index for each field and they map to the ids (which could be miillions to billion) ?

**Reply by xiaofan-luan** [reply]
usually u start from filter with no index and that could be fast enough for many use cases.

If it's not, then usually create an inverted index is what we recommended. 

you can use mmap mode to map index or raw data into disk to save memory

**Comment by patelprateek** [comment]
@yhmo Could you please help me understand how the pre-filtering works.
HNSW has its own internal id to embedding mapping and then the other index have there own perhaps internal mapping , once we get pre-filtered documents , do we translate the ids of the documents to match the internal id of hnsw  ? Will this scale to say cases where we have pre-filtering resulting in millions o ids ?

**Reply by xiaofan-luan** [reply]
can you explain about your use case?
the internal data structure of Milvus is complicated. we split data into segments , each segment is usually around 1GB and 1million data.

When filter happens, milvus first filter with expression and generate internal ID, then covert the ID into a bitset and HNSW id is mapping to the internal ID

**Reply by yhmo** [reply]
@patelprateek  
Answered in this thread: https://github.com/milvus-io/milvus/discussions/36940
